test/base-tlsutility.cpp

```diff
@@ -85,71 +46,230 @@ BOOST_AUTO_TEST_CASE(sha1)
 
-	BOOST_CHECK(!IsCaUptodate(MakeCert("Icinga CA", key, "Icinga CA", key, [](ASN1_TIME* notBefore, ASN1_TIME* notAfter) {
-		BOOST_REQUIRE(X509_gmtime_adj(notBefore, 0));
-		BOOST_REQUIRE(X509_gmtime_adj(notAfter, LEAF_VALID_FOR - 60 * 60));
-	}).get()));
+	EVP_PKEY *pkey = EVP_PKEY_new();
+	EVP_PKEY_assign_RSA(pkey, rsa);
+	BOOST_CHECK_EQUAL(EVP_PKEY_RSA, EVP_PKEY_id(pkey));
+
+	return {pkey, EVP_PKEY_free};
 }
 
-BOOST_AUTO_TEST_CASE(iscertuptodate_ok)
+/**
+ * Creates a new X509 certificate signed by the Icinga CA based on an existing CA certificate.
+ *
+ * It uses the public key and subject name from the existing CA certificate to create a new certificate.
+ *
+ * @param ca The existing CA certificate.
+ * @param validFor The validity period in seconds.
+ * @param validFrom The start time offset in seconds.
+ * @param caCert Whether to create a CA certificate (true) or a leaf certificate (false).
+ * @returns The new X509 certificate or an empty shared_ptr on error.
+ */
+auto NewCertFromExisting(const std::shared_ptr<X509>& ca, long validFrom, long validFor, bool caCert = false)
 {
-	BOOST_CHECK(IsCertUptodate(MakeCert("Icinga CA", GenKeypair(), "example.com", GenKeypair(), [](ASN1_TIME* notBefore, ASN1_TIME* notAfter) {
-		time_t epoch = 0;
-		BOOST_REQUIRE(X509_time_adj(notBefore, l_2017, &epoch));
-		BOOST_REQUIRE(X509_gmtime_adj(notAfter, RENEW_THRESHOLD + 60 * 60));
-	})));
+	std::shared_ptr<EVP_PKEY> caPubKey(X509_get_pubkey(ca.get()), EVP_PKEY_free);
+	return CreateCertIcingaCA(caPubKey.get(), X509_get_subject_name(ca.get()), validFrom, validFor, caCert);
 }
 
-BOOST_AUTO_TEST_CASE(iscertuptodate_expiring)
+/**
+ * Creates an ASN1_TIME object representing the current time plus the specified number of seconds.
+ *
+ * @param seconds The number of seconds to add to the current time.
+ * @returns A shared pointer to the ASN1_TIME object.
+ */
+static std::shared_ptr<ASN1_TIME> MakeASN1TimeFrom(long seconds)
 {
-	BOOST_CHECK(!IsCertUptodate(MakeCert("Icinga CA", GenKeypair(), "example.com", GenKeypair(), [](ASN1_TIME* notBefore, ASN1_TIME* notAfter) {
-		time_t epoch = 0;
-		BOOST_REQUIRE(X509_time_adj(notBefore, l_2017, &epoch));
-		BOOST_REQUIRE(X509_gmtime_adj(notAfter, RENEW_THRESHOLD - 60 * 60));
-	})));
+	auto now = time(nullptr);
+	return {X509_time_adj_ex(nullptr, 0, seconds, &now), ASN1_TIME_free};
 }
 
-BOOST_AUTO_TEST_CASE(iscertuptodate_old)
+/**
+ * Formats an ASN1_TIME object as a human-readable string.
+ *
+ * @param t The ASN1_TIME object to format.
+ *
+ * @returns A string representation of the ASN1_TIME object.
+ */
+static std::string FormatAsn1Time(const ASN1_TIME* t)
 {
-	BOOST_CHECK(!IsCertUptodate(MakeCert("Icinga CA", GenKeypair(), "example.com", GenKeypair(), [](ASN1_TIME* notBefore, ASN1_TIME* notAfter) {
-		time_t epoch = 0;
-		BOOST_REQUIRE(X509_time_adj(notBefore, l_2016, &epoch));
-		BOOST_REQUIRE(X509_gmtime_adj(notAfter, RENEW_THRESHOLD + 60 * 60));
-	})));
+	BIO *bio = BIO_new(BIO_s_mem());
```

Comment:
This could use the `shared_ptr` trick to automatically free this one as well as commonly used elsewhere in this file.